db/python/tables/bq/billing_base.py

```diff
@@ -0,0 +1,673 @@
+import re
+from abc import ABCMeta, abstractmethod
+from collections import Counter, defaultdict, namedtuple
+from datetime import datetime
+from typing import Any
+
+from google.cloud import bigquery
+
+from api.settings import BQ_BUDGET_VIEW, BQ_DAYS_BACK_OPTIMAL
+from api.utils.dates import get_invoice_month_range, reformat_datetime
+from db.python.gcp_connect import BqDbBase
+from models.models import (
+    BillingColumn,
+    BillingCostBudgetRecord,
+    BillingTimePeriods,
+    BillingTotalCostQueryModel,
+)
+
+# Label added to each Billing Big Query request,
+# so we can track the cost of metamist-api BQ usage
+BQ_LABELS = {'source': 'metamist-api'}
+
+
+# Day Time details used in grouping and parsing formulas
+TimeGroupingDetails = namedtuple('TimeGroupingDetails', ['field', 'formula'])
+
+
+def abbrev_cost_category(cost_category: str) -> str:
+    """abbreviate cost category"""
+    return 'S' if cost_category == 'Cloud Storage' else 'C'
+
+
+def prepare_time_periods(
+    query: BillingTotalCostQueryModel,
+) -> TimeGroupingDetails:
+    """Prepare Time periods grouping and parsing formulas"""
+    time_column = query.time_column or 'day'
+    result = TimeGroupingDetails('', '')
+
+    # Based on specified time period, add the corresponding column
+    if query.time_periods == BillingTimePeriods.DAY:
+        result = TimeGroupingDetails(
+            field=f'FORMAT_DATE("%Y-%m-%d", {time_column}) as day, ',
```

Comment:
Should these have to end with a `, `, or can the usage handle this?

Reply:
Well later it is used to build SQL, by default that time field & formula is empty.
Without that ending ',' the usage would need to do extra check and if not empty append ','

Reply:
I have removed the comma in the end, instead added new attribute called separator, which gets added when SQL is constructed.